transformation/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/ThreadSafeTransformerWrapper.java

```diff
@@ -0,0 +1,132 @@
+ */
+@Slf4j
+public class ThreadSafeTransformerWrapper implements IJsonTransformer {
+
+    public static final Cleaner FALLBACK_CLEANER = Cleaner.create();
+
+    /** Thread-local holding per-thread transformer wrappers */
+    private final ThreadLocal<CloseTrackingTransformer> threadLocalHolder;
+
+    /**
+     * Constructs a new thread-safe transformer wrapper using the given supplier.
+     *
+     * @param transformerSupplier Supplier that produces a new {@link IJsonTransformer} for each thread.
+     */
+    public ThreadSafeTransformerWrapper(Supplier<IJsonTransformer> transformerSupplier) {
+        this.threadLocalHolder = ThreadLocal.withInitial(() -> {
+            var trackingTransformer = new CloseTrackingTransformer(transformerSupplier.get());
+            // Register the tracking wrapper with the Cleaner. The cleaner will call cleanup() when the ThreadSafeTransformerWrapper is garbage collected
+            FALLBACK_CLEANER.register(this, trackingTransformer::cleanup);
+            return trackingTransformer;
+        });
+    }
+
+    @Override
+    public Object transformJson(Object input) {
+        return threadLocalHolder.get().transformJson(input);
+    }
+
+    /**
+     * Manually closes the transformer associated with the current thread.
+     * This must be called prior to any calling thread being shutdown.
+     */
+    @Override
+    public void close() {
+        var transformer = threadLocalHolder.get();
+        if (transformer != null) {
+            transformer.close();
+            threadLocalHolder.remove();
+        }
+    }
+
+    /**
+     * A close-tracking wrapper around an {@code IJsonTransformer} that monitors whether it has been closed.
+     */
+    private static class CloseTrackingTransformer implements IJsonTransformer {
+        private final IJsonTransformer delegate;
+        // Volatile flag for thread-safe check of the closed state.
+        private volatile boolean closed = false;
+
+        CloseTrackingTransformer(IJsonTransformer delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Invoked by the Cleaner if the transformer is garbage-collected without an explicit close().
+         * This checks whether close() was already called, and if not, performs cleanup.
+         */
+        private void cleanup() {
+            if (!closed) {
+                log.atWarn()
+                    .setMessage("Fallback cleaner used to cleanup transformer {}. Explicit close() was not called.")
+                    .addArgument(delegate)
+                    .log();
+                try {
+                    delegate.close();
+                } catch (Exception e) {
+                    log.atError()
+                        .setMessage("Exception during fallback cleaning of transformer {}.")
+                        .addArgument(delegate)
+                        .setCause(e)
+                        .log();
+                }
+                closed = true;
```

Comment:
should this be in a finally block?

Reply:
Should this be an atomic?  Can cleanup be called from any thread?  You're checking closed in the transform method